source/extensions/filters/http/ext_proc/ext_proc.cc

```diff
@@ -24,12 +24,17 @@ using Http::ResponseHeaderMap;
 
 static const std::string kErrorPrefix = "ext_proc error";
 
-void Filter::openStream() {
+Filter::StreamOpenState Filter::openStream() {
   if (!stream_) {
     ENVOY_LOG(debug, "Opening gRPC stream to external processor");
     stream_ = client_->start(*this);
     stats_.streams_started_.inc();
+    if (processing_complete_) {
+      // Stream failed while starting and either onGrpcError or onGrpcClose was already called
+      return sent_immediate_response_ ? StreamOpenState::Error : StreamOpenState::IgnoreError;
```

Comment:
Should this happen earlier, i.e. at the start of this basic block, before we try to reopen the stream again with the `start`?

Reply:
Also, why is it legit to reopen a stream ever? In general, you might end up bouncing to a compeltely different server..

Reply:
No, all the callback functions check this before calling openStream() and bail out early.

We'll only call openStream() the first time the filter needs to send something on the stream, and since we have a flexible processing mode, any callback could be the first one to need to open the stream. The callback actually never resets stream_ (it's a unique_ptr) so we have a failsafe to ensure it's never opened twice.

The bug that this PR fixes is that the gRPC client might call onGrpcError as part of the "client_->start()" call. So, if processing_complete_ is set now, then it means that this happened, which tells us that the stream failed to open.

I think this is all correct. 

One thing I could see doing is asserting that processing_complete_ is false at the start of this method just to ensure that all the other callbacks are calling it appropriately. Either that or come up with some sort of "state" enum for the filter to make all of these lifecycle things more explicit.